Multiscreen, and allow pixel level window specifications.

zelch opened the pull request:
There are a few too many things in this for my preferences, but the changes are tied together in important ways.

Significant changes:

1. You can now specify the X, Y, width, and height as 'raw' values, which are in pixels.
2. The layout QML files are now reloaded every time you bring up the tiling dialog. (This was more of an accident than anything else.  We need to recalculate things when bringing up the dialog on a different screen.)
3. You can now specify a screen number for a given layout. (This is _not_ per window, because I can't come up with a sane way to render that in the dialog.  So people are going to be stuck with a layout put monitor.)
4. And the layout names are now used for something!  We get a text label above each layout in the dialog.
5. There are also some bug fixes around showing vs hiding the dialog, and how that interacts with the running focus task.  This also fixed some bugs around shortcuts with number 2.

Commit:
Pass the window object into tileWindow.

This is instead of extracting values from the object and passing those
as arguments.

In addition, and quite related, the naming in tileWindow now matches
other places in the code, where the client winow is usually named
client, and window is usually used to refer to the window configuration.

Commit:
Add support for raw values in pixels.

That is, if you say 'rawWidth: 1080', then your window would be set to
1080 wide.

If you have both raw values and the existing units of 12 values, the raw
values are used.

All of this allows very precise window geometry to be specified, at the
noticible cost that any configuration given in pixels is going to be
specific to that monitor resolution.

Also note, the coordinates are given for a given desktop and screen,
instead of in the global address space.

The latter would 'solve' the multi-monitor problem, but that would end
up with an entirely different way of thinking about coordinates between
the raw values and the plain values.

Though I do need to make a PR for the Window Geometry Information kwin
script, as that currently only gives window coordinates in the global
coordinate state.

Here is an example window configuration using the new functionality:

        {
            rawY: 651,
            rawWidth: 2537,
            rawHeight: 1439,
            x: 0.00,
            y: 4.01,
            width: 5.94,
            height: 7.99,
            shortcutModifier: Qt.AltModifier,
            shortcutKey: Qt.Key_1
        },

Commit:
Clear shortcuts on hide.

We add the shortcuts to the list every time that we bring up the dialog,
which means that we slowly accumulate a decent number of them, and all
of the associated data structures.

In the more extreme cases, this can absolutely end up with us taking out
kwin entirely with references to stale objects.

Commit:
Use hide().

The hide function does a bit more than just making the dialog
non-visible.

And we want those things done whenever we are hiding the dialog.

There are still cases of direct manipulation of hide in WindowLayout,
but this is a decent start.

Commit:
Do a hide before doing a show.

This is a matter of paranoia, we really want to make sure that things
like the shortcut list have been cleared before we do a show.

Commit:
Start off not running the timer.

We should only be running the timer when the dialog is up on the screen.

As such, we really shouldn't be starting with it on as soon as we load.

Commit:
Break this into multiple lines.

It was needed during debugging.

Commit:
Rework how we do layouts for the dialog.

There are a few different things here, all of them important.

We are bounding the size of the boxes to 12, though we are not bothering
to ensure that the x/y + width/height comes out to 12 or less, so we may
still draw outside of the box, but not as much at least.

We are using Math.round to do a round-nearest on the resulting
x/y/width/height value, ensuring that we're not only dealing with
integers, but with integers that have been correctly rounded.

And we're passing screen into the WindowLayout object, so that we don't
get errors about binding to a non-NOTIFYable variable.

Oh, and a typo fix.

Commit:
Pass mainDialog into WindowLayout as main.

It is annoyingly difficult to pass a var into a sub object with the same
name, so it's main inside WindowLayout.

We're also doing the whole visible = false -> hide() changeover here,
because those lines are changing with the var name change.

Commit:
Ensure that we get the tile state correct.

With key bindings, it seems that when tileWindow is called it does not
always have access to a filled in WindowLayout object.

As such, we need to pass in all the state that we need.

Commit:
Use the correct variables here.

Commit:
Up the QtQuick.Layouts version.

Commit:
Regenerate all the layouts in show().

This is necessary to have things render correctly when moving between
two displays.

As a very significant side bonus, this means that it rereads the layout
files.

So you can now modify a layout file, and just opening the dialog again
will cause the modified version to be used.

Commit:
Put layout names above the layout render.

This makes the name useful.

We should probably make the font size, and possibly even the presence of
the label, configuration settings.

Commit:
Allow a layout to specify a screen number.

Note, this does depend on the screens keeping the same numbers.

We'll see how well that works going forward.

Commit:
Tweak an error message.

qewer33 commented on package/contents/ui/main.qml:
```diff
@@ -51,6 +52,18 @@ PlasmaCore.Dialog {
+        mainDialog.screen = workspace.clientArea(KWin.MaximizeArea, workspace.activeScreen, workspace.currentDesktop);
+        var screen = mainDialog.screen
+
+        // We want layoutsRepeater to regenerate all of it's children.
+        // To make that happen, we save the model it is using, set the model to
+        // undefined, and then restore the model.
+        // It's a hack, but it works.
```
I remember trying this exact hack when I was initially developing Exquisite a year ago but it didn't seem work. Are you certain it works fine?

qewer33 replied:
Tested and it works.

qewer33 commented:
Hey! Thanks a lot for the PR again. It looks good at first glance with no conflicts, I'm going to be adding some reviews since I need to understand what you've changed. Hope you don't mind.

A quick review of some of your points:

>- You can now specify the X, Y, width, and height as 'raw' values, which are in pixels.

This is a very nice addition. Thanks!

>- You can now specify a screen number for a given layout. (This is not per window, because I can't come up with a sane way to render that in the dialog. So people are going to be stuck with a layout put monitor.)

I see the idea you decided to go with here and it makes sense but I was thinking something different for multi-monitor support. [ExquisiteW](https://github.com/imthenachoman/ExquisiteW) implements it like this:

![image](https://github.com/qewer33/Exquisite/assets/69015181/08d3c397-42eb-47e5-a9a6-0bfba1c5a3bc)

So the current monitor would be selected by default and if the user wants to send window to another monitor, they can select it. Would be great to hear your opinion here!

>- And the layout names are now used for something! We get a text label above each layout in the dialog.

This is a great addition too. But it might be a good idea to add a preference for it. You can give it a shot or I can add it myself after this is merged.

>- There are also some bug fixes around showing vs hiding the dialog, and how that interacts with the running focus task. This also fixed some bugs around shortcuts with number 2.

Thanks for these too!

qewer33 commented on package/contents/ui/lib/WindowLayout.qml:
```diff
@@ -123,9 +177,14 @@ PlasmaComponents.Button {
-                        tileShortcuts.set(key, () => {
-                            tileWindow(activeClient, window.x, window.y, window.width, window.height);
-                        });
+                        main.tileShortcuts.set(key, function(workspace, window, tileWindow, root) {
+                            return function() {
+                                if (window == undefined || root == undefined || workspace == undefined || workspace.activeClient == undefined) {
+                                    return;
+                                }
+                                tileWindow(workspace.activeClient, window, root);
+                            }
+                        }(workspace, window, tileWindow, root));
```
I'm not sure I understand what you changed here. Would be great if you can explain. And those checks would be better suited inside `tileWindow`

qewer33 replied:
Also just tested and shortcuts don't seem to work at all.

zelch replied:
That's _really_ odd.  This is while the dialog is open, correct?

It's definitely working inside my Debian12 VM.

(It made life easier.  I can mount the filesystem with sshfs to easily copy over the edited files, ssh over to watch the log output, and logout and log back in every time I need to test a change.)

With that said, if you add a console.log statement just before the tileWindow call at what is currently line 185, and one at the start of the function at line 182, do either show up in the logs?

qewer33 commented on package/contents/ui/main.qml:
```diff
@@ -96,6 +108,7 @@ PlasmaCore.Dialog {
     function hide() {
         focusTimer.running = false;
         mainDialog.visible = false;
+        mainDialog.tileShortcuts.clear();
```
I haven't tested yet but since tile shortcuts are registered on `Component.onCompleted`, they only get registered once Exquisite starts along with Kwin. Wouldn't this make tile shortcuts unusable after hiding Exquisite once?

qewer33 replied:
Just saw your commit descriptions:
> We add the shortcuts to the list every time that we bring up the dialog,
which means that we slowly accumulate a decent number of them, and all
of the associated data structures.
>
> In the more extreme cases, this can absolutely end up with us taking out
kwin entirely with references to stale objects.

As I said above, the shortcuts are added on the `Component.onCompleted` callback of each WindowLayout button. `Component.onCompleted` runs only when the object is first instantiated, which is the start of KWin in Exquisite's case. Hiding and showing the Exquisite dialog doesn't re-instantiate objects.

zelch replied:
It didn't used to re-instantiate objects, but the change up at line 59 of main.qml changes that.

It's also, I believe, why I needed to do the complicated state dance on line 180 of WindowLayout.qml.

Without this change, but with the change to re-instantiate things every time we open the dialog, we end up continually adding to the list of shortcuts.

In the worst case, without some of the guard rails I added in, this results in kwin segfaulting, which wasn't something I expected to be possible with QML. :)

qewer33 commented on package/contents/ui/lib/WindowLayout.qml:
```diff
@@ -96,24 +118,56 @@ PlasmaComponents.Button {
-                Layout.rowSpan: windows[index].width
-                Layout.column: windows[index].x
-                Layout.columnSpan: windows[index].height
+                Layout.column: {
+                    if (windows[index].hasOwnProperty("rawX")) {
+                        let val = Math.min(windows[index].rawX / (screen.width / 12.0), 12.0);
+                        val = Math.round(val);
+                        return val;
+                    } else {
+                        return windows[index].x;
+                    }
+                }
+                Layout.row: {
+                    if (windows[index].hasOwnProperty("rawY")) {
+                        let val = Math.min(windows[index].rawY / (screen.height / 12.0), 12.0);
+                        val = Math.round(val);
+                        return val;
+                    } else {
+                        return windows[index].y;
+                    }
+                }
+                Layout.rowSpan: {
+                    if (windows[index].hasOwnProperty("rawWidth")) {
+                        let val = Math.min(windows[index].rawWidth / (screen.width / 12.0), 12.0)
+                        val = Math.round(val);
+                        return val;
+                    } else {
+                        return windows[index].width;
+                    }
+                }
+                Layout.columnSpan: {
+                    if (windows[index].hasOwnProperty("rawHeight")) {
+                        let val = Math.min(windows[index].rawHeight / (screen.height / 12.0), 12.0);
+                        val = Math.round(val);
+                        return val;
+                    } else {
+                        return windows[index].height;
+                    }
+                }
```
Code duplication here can be condensed to a convenience function.

zelch replied:
Hm, indeed.

zelch replied:
Done.

qewer33 commented on package/contents/ui/lib/WindowLayout.qml:
```diff
@@ -13,26 +13,48 @@ PlasmaComponents.Button {
+        let width = 0.0;
+        let height = 0.0;
+
+        if (window.hasOwnProperty("rawX")) {
+            x = window.rawX;
+        } else {
+            x = Math.round(window.x * xMult) + tileGap;
+        }
+        if (window.hasOwnProperty("rawY")) {
+            y = window.rawY;
+        } else {
+            y = Math.round(window.y * yMult) + tileGap;
+        }
+        if (window.hasOwnProperty("rawWidth")) {
+            width = window.rawWidth;
+        } else {
+            width = Math.round(window.width * xMult) - 2*tileGap;
+        }
+        if (window.hasOwnProperty("rawWidth")) {
+            height = window.rawHeight;
+        } else {
+            height = Math.round(window.height * yMult) - 2*tileGap;
+        }
```
Something like
```js
x = window.rawX ? window.rawX : Math.round(window.x * xMult) + tileGap;
```
or (if nullish coalescing is supported in QML JS, not sure about that)
```js
x = window.rawX ?? Math.round(window.x * xMult) + tileGap;
```
would be more concise here.

zelch replied:
Done.

qewer33 commented on package/contents/ui/main.qml:
```diff
@@ -183,13 +196,40 @@ PlasmaCore.Dialog {
+                        return layout.childHasFocus();
+                    }
+
+                    Text {
+                        id: label
+                        text: layoutFile.item.name
+                        color: "white"
+                        font.pointSize: 15
```
This should be a `PlasmaComponents.Label` instead of `Text`. Also remove the color and font size properties.

qewer33 replied:
A preview with some edits:

![image](https://github.com/qewer33/Exquisite/assets/69015181/6dbb1347-e460-4c95-9fa3-38b752a07c48)

```qml
// Kirigami import at the top of the file
import org.kde.kirigami 2.20 as Kirigami

// Text changed to a Label and moved beloew the layout
PlasmaComponents.Label {
    id: label
    text: layoutFile.item.name
    color: Kirigami.Theme.disabledTextColor
    width: layout.width
    elide: Text.ElideRight
    Layout.alignment: Qt.AlignHCenter
}
```

imthenachoman replied:
This is f***ing beautiful. I love it.

zelch replied:
Oh, I really like it after those changes.

I had to do the color because with just Text, it came out dark on dark in night mode, which was unreadable.

zelch replied:
Oh, I really like it after those changes.

I had to do the color because with just Text, it came out dark on dark in night mode, which was unreadable.

Anyhow, change made, though I made the location of the name a configuration setting. :)

Commit:
Replace a bunch of duplicate code with a function.

Thanks to qewer for the suggestion.

Commit:
Use PlasmaComponents.Label for the name.

This is almost exactly qewer33's suggested change, except with a
configuration setting for those of us who like the name on top instead.

Commit:
Use nullish coalescing for more concise code.

I didn't even know nullish coalescing was a thing, but I think I like
it.

zelch commented:
> I see the idea you decided to go with here and it makes sense but I was thinking something different for multi-monitor support. [ExquisiteW](https://github.com/imthenachoman/ExquisiteW) implements it like this:

...

> So the current monitor would be selected by default and if the user wants to send window to another monitor, they can select it. Would be great to hear your opinion here!

Personally, I would find that approach a lot clunkier.

That's mostly because with that arrangement, you have to click on things to do what you want, where as with mine you can put the focus on a window, hit the shortcut to bring up the dialog, and then hit the shortcut to send that window where you want it, even when that somewhere is another monitor.

And especially with the ability to do stuff with pixels, the coordinates won't even make sense on the wrong monitor. (At least, in my setup, the two monitors are very different resolutions.)

zelch commented:
> I see the idea you decided to go with here and it makes sense but I was thinking something different for multi-monitor support. [ExquisiteW](https://github.com/imthenachoman/ExquisiteW) implements it like this:

...

> So the current monitor would be selected by default and if the user wants to send window to another monitor, they can select it. Would be great to hear your opinion here!

Personally, I would find that approach a lot clunkier.

That's mostly because with that arrangement, you have to click on things to do what you want, where as with mine you can put the focus on a window, hit the shortcut to bring up the dialog, and then hit the shortcut to send that window where you want it, even when that somewhere is another monitor.

And especially with the ability to do stuff with pixels, the coordinates won't even make sense on the wrong monitor. (At least, in my setup, the two monitors are very different resolutions.)

imthenachoman commented:
> Personally, I would find that approach a lot clunkier.

Can I get a screenshot of your example? I'm curious how it differs. Based on your description, it sounds like it's more work but maybe I am misunderstanding.

zelch commented:
> > Personally, I would find that approach a lot clunkier.
> 
> Can I get a screenshot of your example? I'm curious how it differs. Based on your description, it sounds like it's more work but maybe I am misunderstanding.

There's nothing in the dialog, except for the windows being based on the correct monitor resolution.

Instead, you set it in the layout QML, so something like this:

```
Item {
    property string name: "Test VM"
    property int screen: 1
    property var windows: [
        {
            rawX: 0,
            rawY: 470,
            rawWidth: 832,
            rawHeight: 590,
            shortcutModifier: Qt.AltModifier,
            shortcutKey: Qt.Key_4
        },
        {
            rawX: 768,
            rawY: 0,
            rawWidth: 832,
            rawHeight: 590,
            shortcutModifier: Qt.AltModifier,
            shortcutKey: Qt.Key_5
        },
        {
            rawX: 768,
            rawY: 590,
            rawWidth: 832,
            rawHeight: 470,
            shortcutModifier: Qt.AltModifier,
            shortcutKey: Qt.Key_6
        }
    ]
}
```

And that will put windows on screen 1 (it's zero based, so that's the second monitor), at those coordinates.

Hm, I suppose we could more or less implement both, where there is the monitor selection radio boxes, but that value is only used if no screen is specified in the file.

imthenachoman commented:
So you would have per monitor layouts? What happens when you remove one of the monitors -- a laptop that is sometimes docked? 

My thought was that your layouts are always the same, regardless of what monitor you want. All that changes is that sometimes you want to move a window to another monitor

> Hm, I suppose we could more or less implement both, where there is the monitor selection radio boxes, but that value is only used if no screen is specified in the file.

What happens when you want to move a window to a layout but on a different monitor?

zelch commented:
> So you would have per monitor layouts? What happens when you remove one of the monitors -- a laptop that is sometimes docked?

The current code defaults to screen 0 if the specified number doesn't exist at the time.

Easy enough to vary it, but there are not too many good options for that case.

Ignoring it would be another option.
> 
> My thought was that your layouts are always the same, regardless of what monitor you want. All that changes is that sometimes you want to move a window to another monitor

This is true when you're giving the locations in 1/12ths of a screen, but it just doesn't work at all when you're specifying the locations in pixels and your monitors are different resolutions.
> 
> > Hm, I suppose we could more or less implement both, where there is the monitor selection radio boxes, but that value is only used if no screen is specified in the file.
> 
> What happens when you want to move a window to a layout but on a different monitor?

A default 'auto' value could work as well.  But that's only going to be useful for some use cases.

qewer33 commented:
We can just implement both. Doesn't add too much complexity anyway. I'll implement the GUI monitor switcher after this PR is merged.

@zelch we have a bigger problem in this PR. Tiling itself doesn't work... I've spotted the reason though.

zelch commented:
> @zelch we have a bigger problem in this PR. Tiling itself doesn't work... I've spotted the reason though.

Oh?  I'm quite curious.

zelch commented:
Alright, found it!

This is a difference in behavior between Wayland and X11, and a potentially useful one at that.

Specifically, workspace.activeClient is ending up as null on X11 while the menu is up, but on Wayland it remains set to the window which had focus.

That means that I should just need to point stuff at the correct activeClient variable.

I'll test this on both X11 and Wayland as soon as I get a minute, but that may well be tomorrow. :)

Commit:
Fix tiling on X11.

There is a fairly important difference in behavior between X11 and
Wayland.

On Wayland, workspace.activeClient is the window with the most recent
focus.

But on X11, workspace.activeClient is undefined after do the raise
stuff.

As such, we just need to use the activeClient on mainDialog.

Commit:
Consolidate all of the raise logic to a function.

Not that there is currently a ton of logic in it.

Commit:
Attempt to detect if we are on Wayland.

This is a horrible hack.

But a bug showed that on Wayland, workspace.activeClient is useful, but
on X11 that is not the case.

And we want to suppress Raise on Wayland.

zelch commented:
Alright @qewer33 , tiling should work on X11 again. :)

Commit:
Remove an errant space at the start of the file.

With a space before the #, the script won't necessarily be run as a bash
script.

This most definitely causes problems.

zelch commented:
Hey @qewer33 , mostly out of curiosity, where is the PR at this point? :)

Commit:
Reload the configuration when we open the dialog.

This is a giant hack, but it works.

Nothing causes the configChanged signal to be sent when a KWin script
has it's settings changed, which means that even if we try to reload the
configuration settings, we just get the cached values.

But it turns out that we can send the signal ourselves, forcing KWin to
reload all of it's configuration settings.

If we then hook into the signal this way, as opposed to the previous
way, we will be notified when everything has actually been handled
inside KWin, and we can then correctly load any new settings.

The upshot of all of this is: We can now properly load any newly changed
configuration settings.

Commit:
Some unimportant cleanups.

Commit:
Add hideTiledWindowTitlebar

This was referenced by both the .ui and the .qml, but didn't exist in
the main.xml, and thus would always be the default value.

Commit:
Drop the text saying to disable and re-enable.

It is no longer necessary to handle configuration updates, all anyone
needs to do is do a fresh open of the dialog.

Commit:
Remove some non-existent settings.

zelch commented:
Alright, I got really fed up with one more thing, and figured out the evil required to fix it.

The code now also correctly handles updated configuration settings.

Now, if you have the dialog open when you hit Apply, the changes won't take effect until you close it and open it again, but otherwise, we pick up any changes.

No more needing to disable and re-enable the script in order to make a config change!

qewer33 commented:
Sorry I've been kinda busy. I'll take a look the PR again this weekend and possibly merge. Thanks again!

qewer33 commented:
Sorry for putting this on hold for so long. Really glad to see the apply preferences fix working nicely too, thanks a lot for the PR again @zelch!
